### PR TITLE
Update the sample app spec

### PR DESCRIPTION
**Pull request by humivo**
Description: Update the sample app spec with test assertions and other edits.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sample-apps/SampleAppSpec.md [hunk 1]
```diff
@@ -55,7 +55,7 @@ This configuration file should be set by an environment variable (SAMPLE_APP_CON
 ```
 ---
 Host: "0.0.0.0"                       # Host - String Address
-Port: "8080"                          # Port - String Port
+Port: "4567"                          # Port - String Port
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:58]: Why is this change required? From my understanding the only sample app that leverages this port is Java? Should java be changed instead?

Comment (bryan-aguilar) [reply sample-apps/SampleAppSpec.md:58]: Also, the port is configurable so isn't this merely an example?

File: sample-apps/SampleAppSpec.md [hunk 2]
```diff
@@ -143,6 +143,144 @@ These are Key Value pairs to be added on metrics.
 }
 ```
 
+**Expected Metrics Test File**
+
+The Metric names should have "_{{testingID}}" added onto the end of the metric name to ensure the correct metric is grabbed for verification.  The {{Testing ID}} is usually set to the INSTANCE_ID environment variable so an example of a metricName is `cpuUsage + "_" + INSTANCE_ID` or `cpuUsage_c2f47fc1fb1f5bb7`.
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:148]: I think you are leaking implementation details of the testing into the spec file. Questions that I think should be answered. 

Does the spec clearly state which labels must be attached to each metric?
Does the spec clearly state how to add a unique id onto each metric name? 

If the answer to both of those is yes then I do not think we need to show how we perform validations. This information and the mustache file pasted below is not relevant to implementation of the sample apps in my opinion.

File: sample-apps/SampleAppSpec.md [hunk 3]
```diff
@@ -227,6 +365,79 @@ These are Key Value pairs to be added on traces.
...
 
+**Expected Traces Test Files**
+
+The subsegment span names will be tested against and are displayed below.
+
+The outgoing-sampleapp trace *will not be tested* for this iteration as it is out of scope due to no need to test this functionality.  However, if we were to test this sample app chaining then we would have the assertion validate shape of the trace and it's subsegments/spans.  For example:
+E.g. Go → Go or Go → Ruby or Ruby → Go
+
+
+`/outgoing-http-call`
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:376]: I have the same concern that this is leaking implementation details of the test suite into the spec document. 

Does the spec provide clear instructions on how traces must be built so that this data shape will be generated? 

If so then I think this and the below content is not required.

File: sample-apps/SampleAppSpec.md (hunk 1 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:58]: I thought it would be easier to change the sample app listening ports to default to 4567 since that is the default sample app listening port in the test framework.  Although, this port is configurable so the user could always change it and it only really serves as a starting point.  However, when I was also testing on ecs, there was a port conflict with using 8080 and so I thought following the behavior of the Java app would be better as a standard unless there's a specific reason we should use 8080 instead?

File: sample-apps/SampleAppSpec.md (hunk 3 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:376]: I agree that this could be leaking implementations details so I will revise the spec to clearly state what will be tested for to match the trace map data

File: sample-apps/SampleAppSpec.md (hunk 2 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:148]: Ok I agree I will add in how to add the unique id onto each of the metric names and remove the rest of the unnecessary details

File: sample-apps/SampleAppSpec.md [hunk 4]
```diff
@@ -134,15 +134,17 @@ Asynchronous Gauge
...
 ```
 {
 “signal”:     (string)“metric”
-“language”:   (string)<name of language used>
+“language”:   (string)<name of language used. Should be set to the name of the sample app preceeding "-sample-app" for standardization purposes> 
 “metricType”: (string)“random”
 }
 ```
 
+Upon implementation, the Metric names should also have the instance ID appended onto the end of the name like so "\<metric name\>_\<instance ID\>".  The instance ID can be retrieved through the INSTANCE_ID environment variable.  An example of what this would look like: `cpuUsage_a1b2c3d4e5f6g7h8`.
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:146]: Is there a better place to reflect metric naming requirements? I believe this could be worded a bit differently to represent the case where `INSTANCE_ID` does not exist.  
```suggestion
If the `INSTANCE_ID` environment variable exists then the sample app MUST append it to all metric names following an underscore(`_`). Example: `cpuUsage_a1b2c3d4e5f6g7h8`.
```

File: sample-apps/SampleAppSpec.md [hunk 5]
```diff
@@ -99,7 +99,7 @@ These are Key Value pairs to be added on metrics.
 ```
 {
 “signal”:     (string)“metric”
-“language”:   (string)<name of language used>
+“language”:   (string)<name of language used. Should be set to the name of the sample app preceeding "-sample-app" for standardization purposes>
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:102]: Does it make sense to append `-sample-app` to the language attribute? To me the `language` attribute should just reflect the name of the language. Should we prefer `java` over `java-sample-app`?

File: sample-apps/SampleAppSpec.md (hunk 1 above)
Comment (bryan-aguilar) [reply sample-apps/SampleAppSpec.md:58]: I took a look at the rest of the sample apps and I believe that only the `java` one does not implement a `config.yaml`. All of the sample apps have a configuration file present. In this case I think it's find to change the default to match the java sample app.

File: sample-apps/SampleAppSpec.md [hunk 6]
```diff
@@ -134,15 +134,17 @@ Asynchronous Gauge
 
 **Common Attributes for Random based metrics**
 
-These are Key Value pairs to be added on metrics.
+These are Key Value pairs to be added on metrics and will be tested for.
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:137]:
```suggestion
These Key Value pairs MUST be present on metrics.
```
This conveys the same message in a more succint form. Must implies a requirement which means it should be tested for.

File: sample-apps/SampleAppSpec.md [hunk 7]
```diff
@@ -209,21 +211,23 @@ The events that must be present in every sample app are the following in respect
...
 
+These span names will be tested.
+
 The fourth endpoint must create a span that will have potentially two child spans. “invoke-sampleapp” is the case where there are more than 0 sample apps configured to make a call to.
 “leaf” request is the case where there are no sample apps to make a call to.
 
 **Common Attributes for Trace spans**
 
-These are Key Value pairs to be added on traces.
+These are Key Value pairs to be added on traces and will be tested for as well.
```
Comment (bryan-aguilar) [inline sample-apps/SampleAppSpec.md:226]:
```suggestion
These Key Value pairs MUST be present on trace spans.
```

File: sample-apps/SampleAppSpec.md (hunk 4 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:146]: I think this place makes sense to reflect metric naming requirements.  This wording also works in showing the what is required so I will be including this change.

File: sample-apps/SampleAppSpec.md (hunk 5 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:102]: Oh I think I have have worded it incorrectly or not clear enough but I meant exactly that.  So for the java-sample-app, the language should be set to `java` as it is the name of the sample app that precedes `-sample-app`.

File: sample-apps/SampleAppSpec.md (hunk 6 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:137]: I agree and will be changing to this message.

File: sample-apps/SampleAppSpec.md (hunk 7 above)
Comment (humivo) [reply sample-apps/SampleAppSpec.md:226]: This change will be made too.

**Review by bryan-aguilar — APPROVED** [review]
